src/firstboot/firstboot.c

```diff
@@ -582,23 +586,57 @@ static int prompt_root_password(void) {
         return 0;
 }
 
-static int write_root_shadow(const char *path, const struct spwd *p) {
-        _cleanup_fclose_ FILE *f = NULL;
+static int write_root_shadow(const char *shadow_path, const struct spwd *p) {
+        _cleanup_fclose_ FILE *original = NULL, *shadow = NULL;
+        _cleanup_(unlink_and_freep) char *shadow_tmp = NULL;
+        struct spwd *sp = NULL;
         int r;
 
-        assert(path);
-        assert(p);
+        r = fopen_temporary_label("/etc/shadow", shadow_path, &shadow, &shadow_tmp);
+        if (r < 0)
+                return r;
+
+        original = fopen(shadow_path, "re");
+        if (original) {
+
+                r = sync_rights(original, shadow_tmp);
+                if (r < 0)
+                        return r;
+
+                while ((r = fgetspent_sane(original, &sp)) > 0) {
+
+                        if (streq(sp->sp_namp, p->sp_namp))
+                                r = putspent_sane(p, shadow); // Replace existing matching entry.
+                        else
+                                r = putspent_sane(sp, shadow); // Keep other entries unchanged.
+                        if (r < 0)
+                                return r;
+                }
+                if (r < 0)
+                        return r;
 
-        RUN_WITH_UMASK(0777)
-                f = fopen(path, "wex");
-        if (!f)
-                return -errno;
+        } else {
+                if (errno != ENOENT)
+                        return -errno;
+
+                r = fchmod(fileno(shadow), 0000);
```

Comment:
is this portable? iirc debian-style distros have the file owned by a shadow group, and more bits set, no?

Reply:
Yes, it's `root:root` with `0000` on Arch and `root:shadow` with `0640` on Debian.

Reply:
What would be the best way to handle this? sysusers uses this unconditionally at the moment. If we change it here, we should probably change it there as well. On the other hand, if no one ever complained about this in sysusers, maybe it doesn't matter here either?

Reply:
oh, indeed we do it like that in sysusers. Maybe leave it like it is then, and wait until a Debian person complains and provides a patch?

In general we should always reuse access mode/ownership of what we find, but of course this doesn't work if the file doesn't exist yet at all...

Reply:
My guess would be that Debian never complained, because packaging would not have used sysusers in the past, because they wanted to be "init agnostic". Hopefully the usage of sysusers will pick up now. 
The problem is at this point also academic, because, at least until the current stable, Debian does not ship systemd-firstboot. 

Would a build time setting SHADOWGROUP (defaulting to root) and SHADOWPERMISSIONS (defaulting to 0000) used both here and for sysusers be a good solution?

Reply:
I'll leave that to a different PR since it exceeds the scope of this one. It also seems a bit overkill to me until someone actually complains about the current solution.

Reply:
yeh, leave this out for now.